libraries/GCS_MAVLink/GCS_FTP.cpp

```diff
@@ -0,0 +1,604 @@
+            reply.opcode = FTP_OP::Ack;
+        } else {
+            // dispatch the command as needed
+            switch (request.opcode) {
+                case FTP_OP::None:
+                    reply.opcode = FTP_OP::Ack;
+                    break;
+                case FTP_OP::TerminateSession:
+                case FTP_OP::ResetSessions:
+                    // we already handled this, just listed for completion
+                    if (ftp.fd != -1) {
+                        AP::FS().close(ftp.fd);
+                        ftp.fd = -1;
+                    }
+                    ftp.current_session = -1;
+                    reply.opcode = FTP_OP::Ack;
+                    break;
+                case FTP_OP::ListDirectory:
+                    ftp_list_dir(request, reply);
+                    // check the directory
+                    break;
+                case FTP_OP::OpenFileRO:
+                    {
+                        // only allow one file to be open per session
+                        if (ftp.fd != -1) {
+                            ftp_error(reply, FTP_ERROR::Fail);
+                            break;
+                        }
+
+                        // sanity check that our the request looks well formed
+                        const size_t file_name_len = strnlen((char *)request.data, sizeof(request.data));
+                        if ((file_name_len != request.size) || (request.size == 0)) {
+                            ftp_error(reply, FTP_ERROR::InvalidDataSize);
+                            break;
+                        }
+
+                        request.data[sizeof(request.data) - 1] = 0; // ensure the path is null terminated
+
+                        // get the file size
+                        struct stat st;
+                        if (AP::FS().stat((char *)request.data, &st)) {
+                            ftp_error(reply, FTP_ERROR::FailErrno);
+                            break;
+                        }
+                        const size_t file_size = st.st_size;
+
+                        // actually open the file
+                        ftp.fd = AP::FS().open((char *)request.data, 0);
+                        if (ftp.fd == -1) {
+                            ftp_error(reply, FTP_ERROR::FailErrno);
+                            break;
+                        }
+                        ftp.mode = FTP_FILE_MODE::Read;
+
+                        reply.opcode = FTP_OP::Ack;
+                        reply.size = sizeof(uint32_t);
+                        *((int32_t *)reply.data) = (int32_t)file_size;
+                        break;
+                    }
+                case FTP_OP::ReadFile:
+                    {
+                        // must actually be reading a file
+                        if (ftp.fd == -1) {
+                            ftp_error(reply, FTP_ERROR::FileNotFound);
```

Comment:
Wow, inadequate mapping :-)

Reply:
Yeah... I can also be convinced this should just be Fail (as we don't have good error codes), or InvalidSession, which may be better... The list is [here](https://mavlink.io/en/services/ftp.html#error_codes) if you have any suggestions.